### ISSUE TITLE
[Bug]: alpha array-type not working with RGB image in imshow()

### ISSUE THREAD

**Issue by patquem** [matplotlib/matplotlib#26092]
### Bug summary

Hi,
Whereas `alpha = constant` works with RGB image, this is not the case when working with an `array-type` for alpha.
(In my real case, I can only pass standard imshow parameters like `alpha `to the function of the library I use, and not a RGBA array).
Patrick

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt
from skimage.color import gray2rgb

arr = np.random.random((10, 10))
arr_rgb = gray2rgb(arr)

alpha = np.ones_like(arr)
alpha[:5] = 0.2

plt.figure()
plt.tight_layout()
plt.subplot(121)
plt.title("Expected outcome")
plt.imshow(arr, alpha=alpha, cmap='gray')
plt.subplot(122)
plt.title("Actual outcome")
plt.imshow(arr_rgb, alpha=alpha)
plt.show()
```

### Actual outcome

![image](https://github.com/matplotlib/matplotlib/assets/16154687/89322c0b-4ccd-46e1-9f2a-50b5f3511dfa)

### Expected outcome

![image](https://github.com/matplotlib/matplotlib/assets/16154687/4d7fcedf-2c54-4f07-a35d-ad23afeb9df6)

### Operating system

Windows

### Matplotlib Version

3.7.1

### Matplotlib Backend

TkAgg

### Python version

Python 3.10.11

### Installation

pip

**Comment by patquem** [comment]
hello @rcomer,
Thanks for replying.
The library I use (Orix) manipulates complex object and allows to plot data thanks to this function : [`plot_map()`](https://orix.readthedocs.io/en/latest/reference/generated/orix.plot.CrystalMapPlot.plot_map.html)
The plotting is based on a RGB array (issue from a complex object) and I wish to pass `alpha` to highlight some area of my map. I can't act directly on the RGB array.
Patrick

**Comment by tacaswell** [comment]
This is a bug.  `imshow` basically has two orthogonal code paths, one for if the user passes us a ndim==2 array (which goes through color mapping) and one for for if eth user passes us a ndim==3 array (which more-or-less just goes out).

I suspect that when we implemented the "alapha-as-array" we forgot about the RGB(A) case.

**Comment by tacaswell** [comment]
Yeah, the issue is in https://github.com/matplotlib/matplotlib/blob/dbc906abe5b8a8a5667f81a305e042e7e98c13e9/lib/matplotlib/image.py#L553-L576 where we do apply alpha, but are asking for the scalar alpha not the array alpha.

https://github.com/matplotlib/matplotlib/pull/14889 is where this feature came in and where we picked up the explicit call to `_get_scalar_alpha`.


That said, I think the fix is to add logic to https://github.com/matplotlib/matplotlib/blob/dbc906abe5b8a8a5667f81a305e042e7e98c13e9/lib/matplotlib/image.py#L557-L558 where we promote RGB -> RGBA to take into account a possibly array-like alpha.

I suspect that we have the same bug in https://github.com/matplotlib/matplotlib/blob/dbc906abe5b8a8a5667f81a305e042e7e98c13e9/lib/matplotlib/image.py#L580 which is handling the fully un-sampled case and have confirmed we see the same bug in https://github.com/matplotlib/matplotlib/blob/dbc906abe5b8a8a5667f81a305e042e7e98c13e9/lib/matplotlib/image.py#L554-L556 


I am going to label this as good first issue as it is a clear bug (array alpha should work with  RGB input).  I would say this is a medium difficulty because the image processing code is a bit complicated. Most of that complexity is there for a good reason so be prepared to understand a majority of the `_make_image` function., The only dicey thing to work out is what to do in the case of the user passing both and RGBA array _and_ an array alpha:  Do we error, discard one (and warn), or blend them?

The patch:

```diff
diff --git a/lib/matplotlib/image.py b/lib/matplotlib/image.py
index 135934a244..0ca26bdabf 100644
--- a/lib/matplotlib/image.py
+++ b/lib/matplotlib/image.py
@@ -553,9 +553,11 @@ class _ImageBase(martist.Artist, cm.ScalarMappable):
             else:
                 if A.ndim == 2:  # _interpolation_stage == 'rgba'
                     self.norm.autoscale_None(A)
-                    A = self.to_rgba(A)
-                if A.shape[2] == 3:
+                    A = self.to_rgba(A, alpha=self.get_alpha())
+                elif A.shape[2] == 3:
                     A = _rgb_to_rgba(A)
+                    if alpha := self.get_alpha() is not None:
+                        A[:, :, 3] = self.get_alpha()
                 alpha = self._get_scalar_alpha()
                 output_alpha = _resample(  # resample alpha channel
                     self, A[..., 3], out_shape, t, alpha=alpha)
```

fixes at least the obvious cases, but does not address the "what about RGBA" issue and I think will double apply the alpha in scalar cases.

I don't have time today to chase through those details or write tests.

**Comment by greglucas** [comment]
Oh the fun, there is also a precedent that it doesn't for ScalarMappables too :)
https://github.com/matplotlib/matplotlib/blob/72885cc1ebd458d1b197a43daaaacbfaebbe597e/lib/matplotlib/cm.py#L454-L457

**Comment by tacaswell** [comment]
And via https://github.com/matplotlib/matplotlib/blob/3b30f47d83c8bc8a2d4de4d8e234f2eeada41b88/lib/matplotlib/image.py#L560-L564 correctly if you pass an RGBA array and a scalar alpha they get combined so we have precedent for 3 of the 4 options!

```python
import numpy as np
import matplotlib.pyplot as plt
a = np.zeros([2, 2, 4])
a[:, :, 0] = 1
a[:, :, 3] = np.linspace(0, 1, 4).reshape(2, 2)
fig, (ax1, ax2) = plt.subplots(1, 2)
ax1.imshow(a)
ax2.imshow(a, alpha=.5)
plt.show()
```

**Comment by rcomer** [comment]
So we should add one that raises, to complete the set 🤪

More seriously, I think it’s probably more important for `imshow` to be consistent with itself than consistent with other things, in which case it should also blend for  array alpha.  (Though I have not followed exactly what that `_resample` function does so I don’t know how difficult it would be.)

**Comment by dhiganthrao** [comment]
Hello there! I'd like to contribute to this issue. I'm new to contributing to Matplotlib, so please pardon me if I ask any silly questions! I'd just like to clarify what the issue is, and the general sketch of the solution:

When displaying images using Matplotlib, if the provided image is of 2 dimensions (grayscale image) and an `alpha` parameter is provided, the image is plotted with the given `alpha` parameter being used to make certain pixels of the image opaque by a factor: if a scalar value of `alpha` is provided, the entire image is affected. If an array is provided for `alpha`, then each pixel of the image is affected by the corresponding value in the `alpha` array.

However, if the provided image is of 3 dimensions (RGB image) and an array is provided for `alpha`, the image isn't affected in any way. If a scalar value is provided, the entire image is affected as is what happened in the last case.

If the provided image is of 4 dimensions (RGBA), and a value for alpha is provided, no matter what the value is (scalar or array), the image isn't affected: it uses the preexisting alpha array present in the image.

We'd like to change this behavior: the case for grayscale images works well, so we don't need to change that. For cases where the dimensions of the image are > 2, we'd like to add cases covering all bases: if the image is RGB, add cases where the display function can accept array values for alpha, and if the image is RGBA, decide on a way to let the user know that there are two alpha values: one already provided in the image, and one provided by them. 

The way to start would be by examining the code snippets @/tacaswell listed, and making the corresponding changes.

Please pardon me if I've made any mistake in my clarification! Thanks for patiently reading through this.

**Comment by tacaswell** [comment]
The dimensions are either 2 or 3, and in the case of 3, may have length 3 or 4 in the last dimension (we should fail on 4D input!).

The goal is to make RGB and RGBA input work with both scalar alpha and array alpha (like the the color-mapped case) with alpha blending in the RGBA case.

**Comment by NikosNikolaidis02** [comment]
Hello,

I am currently working on this issue.

I would like to ask if we have a problem if the array is RGB and not RGBA? Should we change the `_rgb_to_rgba` function in order to work for numpy-array-type alpha or just add condition on `image.py` on lines 557 to 558 ?

**Comment by stevezhang1999** [comment]
Just want to clarify that if both RGBA input and parameter ``alpha`` is provided for ``imshow()``, should just ignore ``alpha`` or combine them? Ignoring it makes more sense to me.

I' m a little bit of confused about the code below - is the code desired? It looks a bit weird for me that we are passing alpha to ``_resample()`` with separated RGB and alpha channel.
> And via
> 
> https://github.com/matplotlib/matplotlib/blob/3b30f47d83c8bc8a2d4de4d8e234f2eeada41b88/lib/matplotlib/image.py#L560-L564
> 
> if you pass an RGBA array and a scalar alpha they get combined so we have precedent for 3 of the 4 options!
> ```python
> import numpy as np
> import matplotlib.pyplot as plt
> a = np.zeros([2, 2, 4])
> a[:, :, 0] = 1
> a[:, :, 3] = np.linspace(0, 1, 4).reshape(2, 2)
> fig, (ax1, ax2) = plt.subplots(1, 2)
> ax1.imshow(a)
> ax2.imshow(a, alpha=.5)
> plt.show()
> ```

I can fix it or we can wait for new contributors as this one looks quite straightforward after so many discussions?

**Comment by tacaswell** [comment]
I think we should combine them.  

For better or worse, the scalar alpha case is blending so the array case should blend too.  It would be very confusing if those two things were different.  


If the user has an RGBA array they want to replace the alpha channel of, then doing

```
ax2.imshow(a[:, :, :3] alpha=alpha_array)
```

is possible (either choice we make), but if they want to blend then is a bit more awkward if we choose to unconditionally replace the alpha channel.

If this is interesting to you, please go ahead and do it @stevezhang1999 !

**Comment by stevezhang1999** [comment]
> I think we should combine them.
> 
> For better or worse, the scalar alpha case is blending so the array case should blend too. It would be very confusing if those two things were different.
> 
> If the user has an RGBA array they want to replace the alpha channel of, then doing
> 
> ```
> ax2.imshow(a[:, :, :3] alpha=alpha_array)
> ```
> 
> is possible (either choice we make), but if they want to blend then is a bit more awkward if we choose to unconditionally replace the alpha channel.
> 
> If this is interesting to you, please go ahead and do it @stevezhang1999 !

Is it common that we want to apply another alpha blending on a RGBA images? Now this looks less weird to me, but still a little bit confused about when we want to do it.

**Comment by AALAM98mod100** [comment]
What I remain confused about is combining the alpha provided by the user and in RGBA image. How should I go about addressing this in the PR, need some guidance here. 

As for the `cm.ScalarMappable.to_rbga()` function, I think the support for alpha arrays inherently exist. If a user passes alpha, it will simply get broadcast to the last dimension in the array at the line `xx[:, :, 3] = alpha`. Someone please correct me if I'm wrong on this assumption

```
try:
    if x.ndim == 3:
        if x.shape[2] == 3:
            if alpha is None:
                alpha = 1
            if x.dtype == np.uint8:
                alpha = np.uint8(alpha * 255)
            m, n = x.shape[:2]
            xx = np.empty(shape=(m, n, 4), dtype=x.dtype)
            xx[:, :, :3] = x
            xx[:, :, 3] = alpha
        elif x.shape[2] == 4:
            xx = x
        else:
            raise ValueError("Third dimension must be 3 or 4")
```

**Comment by QuLogic** [comment]
> Is it common that we want to apply another alpha blending on a RGBA images? Now this looks less weird to me, but still a little bit confused about when we want to do it.

We discussed this on call earlier today. I've written a short script to exercise all cases (that we could think of).
![Figure_1](https://github.com/matplotlib/matplotlib/assets/302469/003078cc-5794-4a36-9667-463c6f2c04d3)

<details>
<summary>test script</summary>

```
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
import numpy as np

fig, axs = plt.subplots(3, 4, figsize=(12, 10), layout='compressed')

for ax in axs.flat:
    ax.set(facecolor='red', xticks=[], yticks=[])

mapped = np.array([
    [0.1, 1.0],
    [1.0, 0.1]
])
rgb = np.repeat(mapped[:, :, np.newaxis], 3, axis=2)
rgba = np.concatenate(
    [
        rgb,
        [
            [[1.0], [0.9]],
            [[0.8], [0.7]],
        ]
    ],
    axis=2
)

alpha_scalar = 0.5
alpha_2d = np.full_like(mapped, alpha_scalar)

cmap_with_alpha = ListedColormap(
    np.concatenate([plt.cm.viridis.colors,
                    np.full((len(plt.cm.viridis.colors), 1), alpha_scalar)],
                   axis=1),
)

for ax, alpha, t in zip(axs, [None, alpha_scalar, alpha_2d], ['off', 'float', 'array']):
    ax[0].imshow(mapped, alpha=alpha)
    ax[0].set_title(f'2D, alpha={alpha_scalar} {t}')

    ax[1].imshow(mapped, cmap=cmap_with_alpha, alpha=alpha)
    ax[1].set_title(f'2D with {alpha_scalar} alpha cmap, alpha={alpha_scalar} {t}')

    ax[2].imshow(rgb, alpha=alpha)
    ax[2].set_title(f'RGB, alpha={alpha_scalar} {t}')

    ax[3].imshow(rgba, alpha=alpha)
    ax[3].set_title(f'RGBA, alpha={alpha_scalar} {t}')

plt.show()
```
</details>

The columns are the types of `X` input: colormapped 2D data, colormapped 2D data with a colormap _that includes transparency_, 3D RGB data, 3D RGBA data. The rows are no alpha, scalar alpha, and array alpha (both of 0.5).

In the second row with scalar alpha, we can see that it is applied by multiplying by the data's alpha (or 1 if data didn't have any), i.e, blending the two alphas. This is most clear with RGBA, which is neither equal to RGB-with-0.5-alpha, nor RGBA-without-alpha.

It is the bottom-right two cases that are inconsistent with that, and we agreed on the call that it should be made consistent, and that adding blending for those cases would be best.

> What I remain confused about is combining the alpha provided by the user and in RGBA image. How should I go about addressing this in the PR, need some guidance here.
> What I'm confused about is the changes that need to be made here. Should I pass the full RGBA image in each of the calls to `_resample` and then combine then within the function?

We should try to apply this in a manner that is consistent with the 2D colormapped case. If it does separate resampling of alpha with later combination, then we should do that.

**Comment by AALAM98mod100** [comment]
Thank you for the plots above @QuLogic. They were super helpful. I have implemented the expected plot in #26520 but the alphas seems to be doubly applied when using `imsave()` instead of `imshow()`.

**Comment by stevezhang1999** [comment]
> 

Can you provide a code snippet about how you are using `imsave()` ? I'm not quite sure how this is related to `imsave()`.

**Comment by AALAM98mod100** [comment]
I'm using the test script from QuLogic above as shown 
```py
...
for ax, alpha, t in zip(axs, [None, alpha_scalar, alpha_2d], ['off', 'float', 'array']):
    ax[0].imshow(mapped, alpha=alpha)
    ax[0].set_title(f'2D, alpha={alpha_scalar} {t}')

    ax[1].imshow(mapped, cmap=cmap_with_alpha, alpha=alpha)
    ax[1].set_title(f'2D with {alpha_scalar} alpha cmap, alpha={alpha_scalar} {t}')

    ax[2].imshow(rgb, alpha=alpha)
    ax[2].set_title(f'RGB, alpha={alpha_scalar} {t}')

    ax[3].imshow(rgba, alpha=alpha)
    ax[3].set_title(f'RGBA, alpha={alpha_scalar} {t}')

plt.savefig('imsave.png') # produces erroneous output 
plt.show() # produces correct output but erroneous after saving figure
```
1. The image I get when I do `plt.show()`
    ![image](https://github.com/matplotlib/matplotlib/assets/49747268/ae4cdd8e-fd69-4faf-bf75-2bdbf3ace51b)
2. The image I get when I save it
    ![image](https://github.com/matplotlib/matplotlib/assets/49747268/7e832157-3a69-491d-a59d-ab18ad62fe2b)

**Comment by stevezhang1999** [comment]
So I got you are refering to `imsave` in `savefig`. I will suggest starting from:
- displaying the input of `imsave` to see if it looks correct
- whether your modification is also presented in `imsave` and break something (a quick look tells me it might be `to_rgba`)

**Comment by saikarna913** [comment]
Is this issue still open? I am willing to work on this issue.

**Comment by rcomer** [comment]
The example from https://github.com/matplotlib/matplotlib/issues/26092#issuecomment-1582624675 was fixed by #28437 

![Image](https://github.com/user-attachments/assets/de0135af-fbc8-4c1d-8c75-5eb5eba60f01)

If I've understood correctly, the example from https://github.com/matplotlib/matplotlib/issues/26092#issuecomment-1683341255 is still wrong on the bottom right as of 7e997ae87b

![Image](https://github.com/user-attachments/assets/a73e032e-91f4-484f-b76a-b88656b0a638)

**Comment by rcomer** [comment]
Removed the "has patch" label because the relevant code has been refactored since the patch was written.

**Comment by Timothy-Anders0n** [comment]
Hi everyone,

I'm digging this up again because I've noticed another bug from the case that was supposed to work (the gray scale image).

Original code had an image array of (10, 10). I've noticed that the alpha array on a grayscale image only works until the image array size goes above (nx, ny) for nx > 75 and/or ny > 75.

Using the original poster's code for reproduction:
```python
import numpy as np
import matplotlib.pyplot as plt
from skimage.color import gray2rgb

arr = np.random.random((75, 75))
arr_rgb = gray2rgb(arr)

alpha = np.ones_like(arr)
alpha[:arr.shape[0] // 2] = 0.2

plt.figure()
plt.tight_layout()
plt.subplot(121)
plt.title("Expected outcome")
# plt.imshow(arr, cmap='gray')
plt.imshow(arr, alpha=alpha, cmap='gray')
plt.subplot(122)
plt.title("Actual outcome")
plt.imshow(arr_rgb, alpha=alpha)
plt.show()
```
yields
<img width="540" height="289" alt="Image" src="https://github.com/user-attachments/assets/aec6f25f-af8e-47e7-80bf-591db910696d" />

while
```python
import numpy as np
import matplotlib.pyplot as plt
from skimage.color import gray2rgb

arr = np.random.random((75, 76))
arr_rgb = gray2rgb(arr)

alpha = np.ones_like(arr)
alpha[:arr.shape[0] // 2] = 0.2

plt.figure()
plt.tight_layout()
plt.subplot(121)
plt.title("Expected outcome")
# plt.imshow(arr, cmap='gray')
plt.imshow(arr, alpha=alpha, cmap='gray')
plt.subplot(122)
plt.title("Actual outcome")
plt.imshow(arr_rgb, alpha=alpha)
plt.show()
```
yields
<img width="540" height="286" alt="Image" src="https://github.com/user-attachments/assets/88258e2d-8cfa-4a58-8ebd-3f51617b82d3" />

**Comment by LangQi99** [comment]
> The example from [#26092 (comment)](https://github.com/matplotlib/matplotlib/issues/26092#issuecomment-1582624675) was fixed by [#28437](https://github.com/matplotlib/matplotlib/pull/28437)
> 
> ![Image](https://github.com/user-attachments/assets/de0135af-fbc8-4c1d-8c75-5eb5eba60f01)
> 
> If I've understood correctly, the example from [#26092 (comment)](https://github.com/matplotlib/matplotlib/issues/26092#issuecomment-1683341255) is still wrong on the bottom right as of [7e997ae](https://github.com/matplotlib/matplotlib/commit/7e997ae87b2abd927830445df5d768c835503504)
> 
> ![Image](https://github.com/user-attachments/assets/a73e032e-91f4-484f-b76a-b88656b0a638)

Hi!
I fixed this issue, adjusted the pytest to use RGBA with alpha blending, and added a example.

Here #30523

<img width="1440" height="864" alt="Image" src="https://github.com/user-attachments/assets/af273906-2650-4323-9deb-3895e884b855" />

**Comment by komalkurapati147** [comment]
Hi, I’d like to work on this issue as a beginner contributor.  
I plan to update `imshow()` so that array‑type alpha values are correctly applied to RGB images.  
Please let me know if it’s okay to proceed.

**Comment by ShubhamPatel-16** [comment]
Hi, I’d like to work on this issue as a beginner contributor.
I plan to update imshow() so that array‑type alpha values are correctly applied to RGB images.
Please let me know if it’s okay to proceed

**Comment by melissawm** [comment]
@ShubhamPatel-16 you can go ahead! Please make sure you carefully read our [Contributing guide](https://matplotlib.org/devdocs/devel/index) and the original issue, and let us know if you have questions. Cheers!

**Comment by story645** [comment]
@ShubhamPatel-16 I also suggest you read the closed pull requests to see why they were closed so that you don't try the same solution. Which, a really helpful triage contribution is posting a summary of the failed attempts to this issue.

**Comment by rcomer** [comment]
This issue has two open linked PRs already.  I believe #30795 is correct apart from the changenote.  I have not looked at the more recent PR.

**Comment by timhoffm** [comment]
Let’s just fix the change note and push #30795 through

**Comment by tacaswell** [comment]
Will this be fixed by #31021 ?

**Comment by timhoffm** [comment]
I don't think so, at least it does not touch the same code (the change to `image.py` l.515ff in https://github.com/matplotlib/matplotlib/pull/30795/changes is the relevant fix). I'd be surprised if you could achieve the same effect with code changes in other places.

**Comment by ayshih** [comment]
No, this bug is unrelated to #31021 (because it has to do with how the alpha channel is assembled prior to it being even handed off to the resampler/renderer)
